### PR TITLE
6 using namingstrategy for names resolving

### PR DESCRIPTION
**Pull request by Vendeetta**
Issue: https://github.com/orgs/devastool/projects/1?pane=issue&itemId=71445768

### CODE REVIEW

**Review by andrey-yurzanov — CHANGES_REQUESTED** [review]
Too raw yet

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+public class FactoryContext {
+  /**
+   * Cache of context instances.
+   */
+  Map<Class<?>, Object> CACHE = new HashMap<>();
+
+  /**
+   * Getting instance of naming strategy.
+   *
+   * @param type Class of naming strategy.
```
Comment (Filkoof) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:22]: should we add @param args?

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+   * @return NamingStrategy implementation.
+   */
+  public <T> T getInstance(Class<T> type, Object... args) {
+    T instance = null;
+    if (type != null) {
```
Comment (Filkoof) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:27]: mb better to check it in beginning ?

Like:
```
if (type == null) {
    throw new IllegalArgumentException("Type must not be null");
}
```

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+              .toArray(Class[]::new);
+          instance = type.getConstructor(argsTypes).newInstance(args);
+          CACHE.put(type, instance);
+        } catch (Exception e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    return instance;
+  }
```
Comment (Filkoof) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:43]: It'll be readable if you extract create instance logic to private method. 

```
private <T> T createInstance(Class<T> type, Object[] args) {
    Class<?>[] argsTypes = Arrays.stream(args)
         .map(Object::getClass)
         .toArray(Class[]::new);
    return type.getConstructor(argsTypes).newInstance(args);
 }
```

and then you can just use `CACHE.computeIfAbsent(type, key -> createInstance(key, args))` and return provided value

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        instance = (T) CACHE.get(type);
+      } else {
+        try {
+          var argsTypes = Arrays.stream(args)
+              .map(Object::getClass)
+              .toArray(Class[]::new);
+          instance = type.getConstructor(argsTypes).newInstance(args);
+          CACHE.put(type, instance);
+        } catch (Exception e) {
+          throw new RuntimeException(e);
```
Comment (Filkoof) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:38]: I supose this catch may be more informative. If we look at it from dev view, it's not realy obvious what we catch here. Narrow the abstraction, and we'll understand what we want to handle here. Or we want to handle every possible exception? 

But anyway we can make it more informative if add description:
`throw new RuntimeException("Failed to create an instance of " + type.getName(), e)`

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+ * Context with caching instance.
+ *
+ * @author Evgeniy_Gerasimov
+ * @since 1.0.0
+ */
+public class FactoryContext {
+  /**
+   * Cache of context instances.
+   */
+  Map<Class<?>, Object> CACHE = new HashMap<>();
```
Comment (Filkoof) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:17]: miss modifiers

**Review by Filkoof — DISMISSED** [review]

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/ContextableFactory.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+ * @since 1.0.0
+ */
+public abstract class ContextableFactory {
+  /**
+   * Factory context.
+   */
+  private final FactoryContext context;
+
+  /**
+   * Constructor.
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/ContextableFactory.java:16]: See JavaDocs format here: [EntityColumnDefinition.java](https://github.com/devastool/entity2jooq/blob/7162c1f5e465f1c0d905b6e9d735126992e365f5/entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/EntityColumnDefinition.java#L32)

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntitySchemaDefinitionFactory.java [hunk 7]
```diff
@@ -39,13 +44,15 @@ public class EntitySchemaDefinitionFactory {
   public Optional<EntitySchemaDefinition> build(Class<?> type, Database database) {
     String schemaName;
     Schema schemaAnnotation = type.getAnnotation(Schema.class);
+    Class<? extends NamingStrategy> strategyClass = SnakeCaseStrategy.class;
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntitySchemaDefinitionFactory.java:47]: `Class<? extends NamingStrategy> strategyClass = SnakeCaseStrategy.class;`
I think this is the wrong way. If class doesn't have `Schema` annotation or the field doesn't have `Column` annotation, you can use strategy from `Table` annotation. The default value in the annotation can change at any time, and it's too difficult to change all places where you wrote `SnakeCaseStrategy.class `.

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityTableDefinitionFactory.java [hunk 8]
```diff
@@ -63,7 +64,7 @@ public Optional<EntityTableDefinition> build(Class<?> type, Database database) {
       Optional<EntitySchemaDefinition> schema = schemaFactory.build(type, database);
       if (schema.isPresent()) {
         String name = tableAnnotation.value();
-        NamingStrategy strategy = new SnakeCaseStrategy(); // TODO. Use tableAnnotation.naming()
+        var strategy = (NamingStrategy) getContext().getInstance(tableAnnotation.naming());
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityTableDefinitionFactory.java:67]: `var strategy = (NamingStrategy) getContext().getInstance(tableAnnotation.naming());`
Doesn't work like this?
`var strategy = getContext().getInstance(tableAnnotation.naming());`

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
+package io.github.devastool.entity2jooq.codegen.definition.factory;
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:1]: You need copyright header.
[Copyright example](https://github.com/devastool/entity2jooq/blob/7162c1f5e465f1c0d905b6e9d735126992e365f5/entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/filesystem/ExtFileVisitor.java#L1)

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+   * @param args argument for instance constructor.
+   * @return new instance.
+   */
+  private <T> T createInstance(Class<T> type, Object[] args) {
+    try {
+      var argsTypes = Arrays.stream(args)
+          .map(Object::getClass)
+          .toArray(Class[]::new);
+      return type.getConstructor(argsTypes).newInstance(args);
+    } catch (Exception e) {
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:47]: Maybe `Exception exeception`?

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+   * @param type Class of naming strategy.
+   * @param args argument for instance constructor.
+   * @return NamingStrategy implementation.
+   */
+  public <T> T getInstance(Class<T> type, Object... args) {
+    if (type == null) {
+      throw new IllegalArgumentException("Type cannot be null");
+    }
+
+    return (T) CACHE.computeIfAbsent(type, v -> createInstance(type, args));
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:31]: Maybe `value ->`?

**Review by andrey-yurzanov — CHANGES_REQUESTED** [review]
Some comments

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/ContextableFactory.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+   * Constructs new instance of {@link ContextableFactory}.
+   *
+   * @param context contains custom settings.
+   */
+  public ContextableFactory(FactoryContext context) {
+    this.context = context;
+  }
+
+  /**
+   * Getter
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/ContextableFactory.java:41]: Returns instance of {@link FactoryContext }

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityColumnDefinitionFactory.java [hunk 13]
```diff
@@ -47,8 +53,9 @@ public EntityColumnDefinitionFactory(EntityDataTypeDefinitionFactory typeFactory
    *
    * @param field entity field, annotation {@link Column} is optional
    * @param table meta-information about table
+   * @param type entity class
    */
-  public Optional<EntityColumnDefinition> build(Field field, TableDefinition table) {
+  public Optional<EntityColumnDefinition> build(Field field, TableDefinition table, Class<?> type) {
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityColumnDefinitionFactory.java:58]: To use type is bad way, I think. You should have all information in `TableDefinition` already, so use `table.getNamingStrategy()`.

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityColumnDefinitionFactory.java [hunk 14]
```diff
@@ -67,4 +74,19 @@ public Optional<EntityColumnDefinition> build(Field field, TableDefinition table
         )
     );
   }
+
+  // if columnAnnotation present return naming strategy from it naming parameter, else
+  // return naming strategy of schema annotation if present or table annotation.
+  private NamingStrategy getNamingStrategy(Class<?> type, Column columnAnnotation) {
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/EntityColumnDefinitionFactory.java:80]: `Schema` is optional, just use `TableDefinition` instance.

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java (hunk 11 above)
Comment (andrey-yurzanov) [reply entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:47]: Maybe `value` ->?

**Review by andrey-yurzanov — CHANGES_REQUESTED** [review]
Some comments

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/EntityTableDefinition.java [hunk 15]
```diff
@@ -42,10 +44,16 @@ public class EntityTableDefinition extends AbstractTableDefinition {
...
-      List<ColumnDefinition> columns
+      List<ColumnDefinition> columns,
+      Class<? extends NamingStrategy> strategy
   ) {
     super(schema, name, "");
     this.columns = columns;
+    this.strategy = strategy;
+  }
+
+  public Class<? extends NamingStrategy> getStrategy() {
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/EntityTableDefinition.java:55]: Comments

File: entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+  /**
+   * Create new instance of type Class.
+   *
+   * @param type Class of naming strategy.
+   * @param args argument for instance constructor.
+   * @return new instance.
+   */
+  private <T> T createInstance(Class<T> type, Object[] args) {
+    try {
+      var argsTypes = Arrays.stream(args)
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/main/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContext.java:59]: `.stream(args)` to next line, please

**Review by andrey-yurzanov — CHANGES_REQUESTED** [review]

**Review by Filkoof — DISMISSED** [review]

**Review by Sergio-Konovalov — APPROVED** [review]

File: entity2jooq-codegen/src/test/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContextTest.java [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+/**
+ * Tests of {@link FactoryContext}.
+ *
+ * @author Evgeniy_Gerasimov
+ */
+class FactoryContextTest {
+
```
Comment (andrey-yurzanov) [inline entity2jooq-codegen/src/test/java/io/github/devastool/entity2jooq/codegen/definition/factory/FactoryContextTest.java:29]: unnecessary line

**Review by andrey-yurzanov — CHANGES_REQUESTED** [review]

**Review by andrey-yurzanov — APPROVED** [review]

**Review by Sergio-Konovalov — APPROVED** [review]

**Pull request merged**